### PR TITLE
client: Fix snap protocol encoding decoding as per spec

### PR DESCRIPTION
**Pull request by g11tech**
Match snap protocol's expected encoding/decoding as per the spec.

### CODE REVIEW

File: packages/client/lib/net/protocol/snapprotocol.ts [hunk 1]
```diff
@@ -1,19 +1,30 @@
...
+import type { AccountBodyBuffer } from '@ethereumjs/util'
 
 interface SnapProtocolOptions extends ProtocolOptions {
   /* Blockchain */
   chain: Chain
+  /**
+   * If to convert slim body received of an account to normal while decoding.
+   * Encoding is always converted to slim
+   */
+  convertSlimBody?: boolean
```
Comment (jochem-brouwer) [inline packages/client/lib/net/protocol/snapprotocol.ts:22]: Why would we want to keep the slim body?

File: packages/client/lib/net/protocol/snapprotocol.ts [hunk 2]
```diff
@@ -124,11 +154,12 @@ export class SnapProtocol extends Protocol {
       name: 'GetStorageRanges',
       code: 0x02,
       response: 0x03,
+      // [reqID: P, rootHash: B_32, accountHashes: [B_32], startingHash: B, limitHash: B, responseBytes: P]
       encode: ({ reqId, root, accounts, origin, limit, bytes }: GetStorageRangesOpts) => {
         return [
           bigIntToBuffer(reqId ?? ++this.nextReqId),
-          root,
-          accounts,
+          setLengthLeft(root, 32),
```
Comment (jochem-brouwer) [inline packages/client/lib/net/protocol/snapprotocol.ts:161]: I am honestly a bit surprised that the spec marks this as a `B_32`. We could save some bytes if the hash has leading zeros, why not shave those off?

File: packages/client/lib/net/protocol/snapprotocol.ts [hunk 3]
```diff
@@ -239,6 +286,7 @@ export class SnapProtocol extends Protocol {
     super(options)
 
     this.chain = options.chain
+    this.convertSlimBody = options.convertSlimBody
```
Comment (jochem-brouwer) [inline packages/client/lib/net/protocol/snapprotocol.ts:289]: This defaults to `undefined` / `false`?

**Review by jochem-brouwer — COMMENTED** [review]
Some questions regarding the convert slim body option.

File: packages/client/lib/net/protocol/snapprotocol.ts (hunk 3 above)
Comment (g11tech) [reply packages/client/lib/net/protocol/snapprotocol.ts:289]: yes, i wasn't sure what we want default to be, so chose that this be passed as explicit flag to make any conversions

File: packages/client/lib/net/protocol/snapprotocol.ts (hunk 1 above)
Comment (g11tech) [reply packages/client/lib/net/protocol/snapprotocol.ts:22]: is slim body not useful at all? then we can always convert and remove this flag totally

**Review by jochem-brouwer — APPROVED** [review]
I will approve here, I am not entirely sure if we want to have the slim account feature support and if it defaults to `false` (when decoding), but I think we will quickly learn in the case that we do/do not want it to remove/keep it. This is one of these cases which we will "learn by practice" :)

Thanks @g11tech :smile:

**Comment by holgerd77** [comment]
Just to note, if you guys want fee free to merge in again.

**Pull request merged**